File: bindings/python/examples/read-write.py
#!/usr/bin/env python
#
# Simple script showing how to read and write data to a device in noncontinuous
# mode, use Ctrl-C to exit.

from __future__ import print_function

from signal import signal, SIG_DFL, SIGINT
import sys
import random

from pysmu import Session, Mode


# If stdout is a terminal continuously overwrite a single line, otherwise
# output each line individually.
if sys.stdout.isatty():
    output = lambda s: sys.stdout.write("\r" + s)
else:
    output = print


def refill_data(num_samples, v=None):
    if v is None:
        # fill channels with a static, random integer between 0 and 5
        v = random.randint(0, 5)
    return [v] * num_samples
    # fill channels with a static, random float between -0.2 and 0.2
    #return [random.uniform(-0.2,0.2)] * num_samples


if __name__ == '__main__':
    # don't throw KeyboardInterrupt on Ctrl-C
    signal(SIGINT, SIG_DFL)

    session = Session()

    if session.devices:
        # Grab the first device from the session.
        dev = session.devices[0]

        # Set both channels to source voltage, measure current mode.
        chan_a = dev.channels['A']
        chan_b = dev.channels['B']
        chan_a.mode = Mode.SVMI
        chan_b.mode = Mode.SVMI
        #chan_a.mode = Mode.SIMV
        #chan_b.mode = Mode.SIMV

        i = 0
        num_samples = 1024

        while True:
            # If stdout is a terminal change the written value every ten
            # iterations, otherwise change it on every iteration.
            if sys.stdout.isatty():
                v = i / 10
            else:
                v = i

            # Write iterating voltage values to both channels.
            chan_a = refill_data(num_samples, v % 6)
            chan_b = refill_data(num_samples, v % 6)
            session.write([[chan_a, chan_b]])
            i += 1

            # Read incoming samples from the first device in a blocking fashion.
            samples = session.get_samples(num_samples)[0]
            for x in samples:
                output("{: 6f} {: 6f} {: 6f} {: 6f}".format(x[0][0], x[0][1], x[1][0], x[1][1]))
    else:
        print('no devices attached')

File: examples/read-write.cpp
// Simple example for reading/writing data in a non-continuous fashion.

#ifdef _WIN32
#include <io.h>
#else
#include <unistd.h>
#endif

#include <array>
#include <chrono>
#include <functional>
#include <iostream>
#include <vector>
#include <system_error>
#include <thread>

#include <libsmu/libsmu.hpp>

#ifndef _WIN32
int (*_isatty)(int) = &isatty;
#endif

using std::cout;
using std::cerr;
using std::endl;

using namespace smu;

int main(int argc, char **argv)
{
	// Create session object and add all compatible devices them to the
	// session. Note that this currently doesn't handle returned errors.
	Session* session = new Session();
	session->add_all();

	if (session->m_devices.size() == 0) {
		cerr << "Plug in a device." << endl;
		exit(1);
	}

	// Grab the first device from the session.
	auto dev = *(session->m_devices.begin());

	// Set device channels to source voltage and measure current.
	dev->set_mode(0, SVMI);
	dev->set_mode(1, SVMI);

	std::vector<std::array<float, 4>> rxbuf;
	std::vector<float> a_txbuf;
	std::vector<float> b_txbuf;

	// refill Tx buffers with data
	auto refill_data = [=](std::vector<float>& buf, unsigned size, int voltage) {
		buf.clear();
		for (unsigned i = 0; i < size; i++) {
			buf.push_back(voltage);
		}
	};

	uint64_t i = 0;
	uint64_t v = 0;

	while (true) {
		// If stdout is a terminal change the written value approximately
		// every second, otherwise change it on every iteration.
		if (_isatty(1))
			v = i / 100;
		else
			v = i;

		// Refill outgoing data buffers.
		refill_data(a_txbuf, 1024, v % 6);
		refill_data(b_txbuf, 1024, v % 6);
		i++;

		try {
			// Write iterating voltage values to both channels.
			dev->write(a_txbuf, 0);
			dev->write(b_txbuf, 1);

			// Run the session for 1024 samples.
			session->run(1024);

			// Read incoming samples in a blocking fashion.
			dev->read(rxbuf, 1024, -1);
		} catch (const std::system_error& e) {
			if (!_isatty(1)) {
				// Exit on dropped samples when stdout isn't a terminal.
				cerr << "sample(s) dropped: " << e.what() << endl;
				exit(1);
			}
		}

		for (auto i: rxbuf) {
			// Overwrite a singular line if stdout is a terminal, otherwise
			// output line by line.
			if (_isatty(1))
				printf("\r% 6f % 6f % 6f % 6f", i[0], i[1], i[2], i[3]);
			else
				printf("% 6f % 6f % 6f % 6f\n", i[0], i[1], i[2], i[3]);
		}
	};
}
